File: Backtracking/CombinationSum.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function(candidates, target) {
    let combination = []
    let currentCombination = []

    const helper = (i, sum) => {

        if(sum == target){
            combination.push([...currentCombination])
            return // we are done with this combination so we break out of it
        }

        if(i >= candidates.length || sum > target){
            return 
        }
        
        currentCombination.push(candidates[i])
        helper(i, sum + candidates[i]) //dont increment i yet since we may not be done with number to add.
        currentCombination.pop()

        helper(i + 1, sum) // not including the number in this call so we just increment i and dont add candidates[i] to sum
    } 

    helper(0, 0)

    return combination
};

File: AdvancedAlgorithms/Backtracking/Combinations/CombinationSum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        res, curComb = [], []

        def dfs(i, total):
            if total == target:
                res.append(curComb.copy())
                return
            
            if i >= len(candidates) or total > target: 
                return

            curComb.append(candidates[i])
            dfs(i, total + candidates[i])
            curComb.pop()

            dfs(i + 1, total)


        dfs(0, 0)

        return res